### DISCUSSION TITLE
Convert a NetworkX graph into a PyTorch Geometric Heterogeneous Graph

### DISCUSSION THREAD

**Discussion by negarmaleki96** [pyg-team/pytorch_geometric#4457] [Q&A]
Hi,
I created a heterogeneous directed graph in NetworkX, and now, for applying GNN models, I think I should convert it into a PyTorch Geometric Heterogeneous Graph. I used "pyg_graph = from_networkx(myG)", but as nodes' attributes are not the same, I got the following error: "ValueError: Not all nodes contain the same attributes"

How can I convert my NetworkX graph into a PyTorch Geometric Heterogeneous Graph? (Also, I have to mention that my attributes are a mix of text and numeric values)

**Comment by rusty1s** [comment]
The data format of heterogeneous graphs in PyG and networkx differ, so it is difficult to implement a general solution to convert a networkx graph into a PyG heterogeneous graph. I personally think the best solution is to do the conversion manually (happy to take any feedback though on how you think this can be achieved). For your text data, you would need to embed them first, e.g., via the `SentenceTransformer` library.

**Reply by negarmaleki96** [reply]
Thank you for your response. So, if I convert the heterogeneous graph from networkX into PyG manually, can I use GNN models on top of it? I mean, do the PyTorch GNN models work with a heterogeneous graph where each node and edge has different attributes?

**Reply by rusty1s** [reply]
Yes, that works. Our [heterogenous graph tutorial](https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html) (hopefully) explains it in detail.

**Reply by negarmaleki96** [reply]
I converted my heterogeneous graph into a homogeneous graph, and then I used [from_networkx](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/utils/convert.html) function to convert it to PyG. 

When I used data.ndata, I have got the following error:

AttributeError: 'GlobalStorage' object has no attribute 'ndata'

Could you please help me with this error?

**Reply by rusty1s** [reply]
There does not exist a `ndata` attribute by default. In PyG, you can access attributes directly, e.g., `data.x` or `data_edge_index`.
